topics/data-science/tutorials/python-exceptions/tutorial.md

```diff
@@ -0,0 +1,163 @@
+{: .tip}
+
+Now we get a much more useful error message from the function! Using `raise` is especially important for *library authors*, people writing python modules that we all use. If they provide useful error messages, it helps you as an end user understand what's happening.
+
+## Try & Except
+
+Let's look at how you would handle one of these exceptions, we'll continue with the `mean()` example above.
+
+```python
+numbers = []
+try:
+    result = mean(numbers)
+    print(f"Result is {result}")
+except ValueError:
+    print("We cannot calculate the mean")
+```
+
+Here we use `try:` to setup a new block, and this code is **tried**, Python attempts to execute it. Below are one or more `except:` blocks which catch specific errors. Here we have specifically said we know a ValueError can happen, and decided to handle it.
+
+Or for another example, accessing a user supplied file. Oftentimes users will call your program and supply a non-existent, or inacessible file. Here you can use multiple `except` blocks to catch all of those potential errors.
+
+```python
+user_supplied_file = 'does-not-exist.txt'
+try:
+    open(user_supplied_file, 'r')
+except FileNotFoundError:
+    print(f"The path you supplied ({user_supplied_file}) doesn't exist, please double check it!")
+except PermissionError:
+    print(f"You supplied a valid file, but it is unreadable. Try changing it's permissions with `chmod +r {user_supplied_file}`")
+```
+
+Failing to open a file raises a `FileNotFoundError` which indicates the file isn't available, and `PermissionError` indicates that a file is unreadable. However in practice, sometimes you'll see something like this:
+
+```
+# Bad!
+try:
+    doSomething()
+except:
+    print("error")
+```
+
+This is called a **bare exception**, and will catch any exception, compared with `except ValueError` which only catches value errors. People consider this generally a bad idea, termed *code smell*. (Because it smells (appears) bad!)
+
+## Finally
+
+The last portion of the `try:`/`except:` block is `finally:`, a third block which lets us do cleanup. It's often very nice to your users that if your program fails halfway through, that you cleanup after yourself.
+
+```python
+import os
+
+try:
+    with open('gene_query.fa', 'w') as handle:
+        handle.write(">some fasta sequence we want to search against a database")
+
+    # runQuery('gene_query.fa')
+
+    # But here we have an error! Something goes wrong!
+    1 / 0
+except:
+    # And now our results are invalid.
+    print("Something went wrong! Oh no! Check your inputs.")
+finally:
+    # So we should cleanup this temporary file we created, so it doesn't cause
+    # problems or distract the user from the results file.
+    os.unlink('gene_query.fa')
```

Comment:
are you now removing the file? or could use as alternative os.remove?
what does os.unlink do exaclty?

should we explain the difference to students?
https://stackoverflow.com/questions/42636018/python-difference-between-os-remove-and-os-unlink-and-which-one-to-use